eudemo/eudemo/equatorial_port.py

```diff
@@ -0,0 +1,200 @@
+        x_o = self.params.x_ob.value
+        z_h = self.params.ep_height.value
+
+        x = (x_i, x_o, x_o, x_i)
+        z = (-z_h / 2, -z_h / 2, z_h / 2, z_h / 2)
+
+        ep_boundary = BluemiraFace(
+            make_polygon({"x": x, "y": 0, "z": z}, label="koz", closed=True)
+        )
+        return ep_boundary
+
+
+class EquatorialPortBuilder(Builder):
+    """
+    Equatorial Port Builder
+    """
+
+    NAME = "Equatorial Port"
+    KOZ = "koz"
+    param_cls: Type[EquatorialPortBuilderParams] = EquatorialPortBuilderParams
+
+    def __init__(
+        self,
+        params: Union[Dict, ParameterFrame, EquatorialPortBuilderParams],
+        build_config: Union[Dict, None],
+        xz_profile: BluemiraFace,
+    ):
+        super().__init__(params, build_config)
+        self.xz_profile = xz_profile
+
+    def build(self) -> Component:
+        """
+        Build the equatorial port component.
+        """
+
+        # TODO: Integrate corner radii
+        self.r_rad = self.params.ep_r_corner.value
+        self.coords = self.params.ep_coordinates.value
+        self.n_ep = self.params.n_EP.value
+
+        self.x_points, self.y_points, self.z_points = castellation(self.coords)
+
+        return self.component_tree(
+            xz=self.build_xz(self.xz_profile),
+            xy=self.build_xy(self.x_points, self.y_points, self.n_ep),
+            xyz=self.build_xyz(self.coords, self.n_ep),
+        )
+
+    def build_xz(self, xz_profile: BluemiraWire) -> PhysicalComponent:
+        """
+        Build the x-z components of the equatorial port
+        """
+
+        body = PhysicalComponent(self.KOZ, xz_profile)
+        body.plot_options.face_options["color"] = BLUE_PALETTE["VV"][0]
+        return body
+
+    def build_xy(self, x: list, y: list, n_ep: int = 10) -> PhysicalComponent:
+        """
+        Build the x-y components of the equatorial port
+        """
+
+        ep_boundary = BluemiraWire(make_polygon({"x": x, "y": y, "z": 0}, closed=True))
+        body = PhysicalComponent(self.NAME, BluemiraFace(ep_boundary))
+        body.plot_options.face_options["color"] = BLUE_PALETTE["VV"][0]
+        return circular_pattern(body, n_shapes=n_ep)
+
+    def build_xyz(
+        self, coords: Union[list, ndarray], n_ep: int = 10
+    ) -> PhysicalComponent:
+        """
+        Build the x-y-z components of the equatorial port
+        """
+        sections = []
+        x_ob = coords[0][-1]  # Outboard x = the last listed x-value
+        new_coords = transpose(coords)
+
+        for r in new_coords:
+            x = r[0]
+            y = [r[1], r[1], r[2], r[2]]
+            z = [r[3], r[4], r[4], r[3]]
+            ext_vec = (x_ob - x, 0, 0)
+            section = BluemiraFace(make_polygon({"x": x, "y": y, "z": z}, closed=True))
+            sections.append(extrude_shape(section, ext_vec))
+
+        ep_shape = boolean_fuse(sections)
+        return circular_pattern(ep_shape, n_shapes=n_ep)

```

Comment:
circular_pattern_component makes this easier as we can just specify the number of TF coils as an input parameter see this function: https://github.com/Fusion-Power-Plant-Framework/bluemira/blob/ad0ee0bf19bb4e7a1f934a6c7e6b5a63d64f8969/bluemira/builders/divertor.py#L90